core/src/main/scala/kafka/coordinator/transaction/TransactionCoordinator.scala

```diff
@@ -364,9 +365,14 @@ class TransactionCoordinator(brokerId: Int,
           val coordinatorEpoch = epochAndTxnMetadata.coordinatorEpoch
 
           txnMetadata.inLock {
-            if (txnMetadata.producerId != producerId)
+            // If we are ending the transaction due to a fenced producer and we are rotating the producer ID because the
+            // epoch was exhausted, we need to skip the typical ID and epoch checks because the transaction metadata
+            // object will already have been changed.
+            val isFencedProducerRotation = txnMetadata.isProducerEpochExhausted &&
+              txnMetadata.pendingState.contains(PrepareEpochFence)
+            if (txnMetadata.producerId != producerId && !isFencedProducerRotation)
               Left(Errors.INVALID_PRODUCER_ID_MAPPING)
-            else if (producerEpoch < txnMetadata.producerEpoch)
+            else if (producerEpoch < txnMetadata.producerEpoch && !isFencedProducerRotation)
               Left(Errors.INVALID_PRODUCER_EPOCH)
             else if (txnMetadata.pendingTransitionInProgress && txnMetadata.pendingState.get != PrepareEpochFence)
```

Comment:
Is this check equivalent to `txnMetadata.pendingState.contains(PrepareEpochFence)`?